api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -414,42 +414,71 @@ public int hashCode() {
 
   public static class NestedField implements Serializable {
     public static NestedField optional(int id, String name, Type type) {
-      return new NestedField(true, id, name, type, null);
+      return new NestedField(true, id, name, type, null, null);
     }
 
     public static NestedField optional(int id, String name, Type type, String doc) {
-      return new NestedField(true, id, name, type, doc);
+      return new NestedField(true, id, name, type, null, doc);
+    }
+
+    public static NestedField optional(int id, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(true, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField optionalWithDefault(int id, String name, Type type, Object defaultValue) {
+      return new NestedField(true, id, name, type, defaultValue, null);
     }
 
     public static NestedField required(int id, String name, Type type) {
-      return new NestedField(false, id, name, type, null);
+      return new NestedField(false, id, name, type, null, null);
     }
 
     public static NestedField required(int id, String name, Type type, String doc) {
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, null, doc);
+    }
+
+    public static NestedField required(int id, String name, Type type, Object defaultValue, String doc) {
+      validateDefaultValueForRequiredField(defaultValue);
+      return new NestedField(false, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField requiredWithDefault(int id, String name, Type type, Object defaultValue) {
+      validateDefaultValueForRequiredField(defaultValue);
+      return new NestedField(false, id, name, type, defaultValue, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type) {
-      return new NestedField(isOptional, id, name, type, null);
+      return new NestedField(isOptional, id, name, type, null, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type, String doc) {
-      return new NestedField(isOptional, id, name, type, doc);
+      return new NestedField(isOptional, id, name, type, null, doc);
     }
 
+    public static NestedField of(int id, boolean isOptional, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(isOptional, id, name, type, defaultValue, doc);
+    }
+
+    private static void validateDefaultValueForRequiredField(Object defaultValue) {
+      Preconditions.checkNotNull(defaultValue, "defaultValue cannot be null for a required field");
+    }
+
+
     private final boolean isOptional;
     private final int id;
     private final String name;
     private final Type type;
+    private final Object defaultValue;
```

Comment:
Are default values supported for container types like struct/map/list?

Reply:
Avro 1.8 supports default values for record, map, array, and enum
from https://avro.apache.org/docs/1.8.1/spec.html#schema_record
```
avro type | json type | example
-- | -- | --
...
record | object | {"a": 1}
enum | string | "FOO"
array | array | [1]
map | object | {"a": 1}
```